packages/Modal/src/Modal.js

```diff
@@ -0,0 +1,149 @@
+
+  /** Exposes a reference to the Modal's portal element. */
+  portalRef: PropTypes.func,
+
+  /** Makes the content of the modal scrollable, but as a side effect prevents any content (e.g. MultiSelects) from
+  overflowing. If you need this on, you may want to consider using a different component instead of this one. */
+  isScrollable: PropTypes.bool,
+
+  /** Width of the Modal. */
+  width: PropTypes.oneOf(ShirtSizes.DEFAULT),
+
+  /** Optional override of z-index. */
+  zIndex: PropTypes.number,
+};
+
+const defaultProps = {
+  className: "",
+  hasOverlay: true,
+  isOpen: false,
+  isScrollable: false,
+  onEscKey: () => {},
+  onOverlayClick: () => {},
+  portalRef: () => {},
+  width: "medium",
+  zIndex: null,
+};
+
+let modalRoot;
+
+class Modal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.modalContainer = null;
+    this.ariaId = uuid();
+
+    // TODO: make this optional
+    if (!modalRoot) {
+      modalRoot = document.createElement("div");
+      document.body.appendChild(modalRoot);
+    }
+
+    this.el = document.createElement("div");
+  }
+
+  componentDidMount() {
+    document.addEventListener("keydown", this.handleKeyDown, false);
+    modalRoot.appendChild(this.el);
+  }
+
+  componentDidUpdate() {
+    this.renderModal();
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener("keydown", this.handleKeyDown);
+    modalRoot.removeChild(this.el);
+  }
+
+  setRef = node => {
+    this.portalRef = node;
+    this.props.portalRef(node);
+  };
+
+  handleKeyDown = event => {
+    if (event.key === "Escape") this.props.onEscKey();
+  };
+
+  renderOverlay() {
+    const { hasOverlay, onOverlayClick } = this.props;
+
+    if (!hasOverlay) return null;
+
+    return <div css={overlayStyles} onClick={onOverlayClick} onKeyDown={onOverlayClick} tabIndex="-1" />;
+  }
+
+  renderModal() {
+    const { ariaLabel, children, className, isOpen, isScrollable, width, zIndex } = this.props;
+    const rootClasses = classNames(className, { "modal--is-open": isOpen });
+    return (
+      <div
+        className={rootClasses}
+        css={modalStyles}
+        isOpen={isOpen}
+        width={width}
+        onKeyDown={this.handleKeyDown}
+        ref={this.setRef}
+        zIndex={zIndex}
+      >
+        {this.renderOverlay()}
+        <div css={frameStyles} width={width}>
+          <section
+            css={contentStyles}
+            isOpen={isOpen}
+            isScrollable={isScrollable}
+            role="dialog"
+            aria-labelledby={this.ariaId}
+          >
+            <span css={stylers.visuallyHidden} id={this.ariaId}>
+              {ariaLabel}
+            </span>
+            {children}
+          </section>
+        </div>
+      </div>
+    );
+  }
+
+  render() {
+    return ReactDOM.createPortal(this.renderModal(), this.el);
+  }
+}
+
+Modal.displayName = "Modal";
+Modal.propTypes = propTypes;
+Modal.defaultProps = defaultProps;
+
+export default Modal;
```

Comment:
Why don't we use some ↪️ in here? 😏 

(may contain some bugs) 
```js
function useEscHandler(handleKeyDown) {
  React.useEffect(() => {
    function handleKeyDown(event) {
      if (event.key === "Escape") {
        handleKeyDown();
      }
    }

    document.addEventListener("keydown", handleKeyDown, false);

    return function cleanup() {
      document.removeEventListener("keydown", handleKeyDown);
    };
  }, [handleKeyDown]);
}

function useModalDiv() {
  const modalRoot = React.useRef(null);
  const el = React.useRef(null);

  React.useEffect(() => {
    modalRoot.current = document.createElement("div");
    document.body.appendChild(modalRoot.current);

    el.current = document.createElement("div");
    modalRoot.current.appendChild(el.current);

    return function cleanup() {
      modalRoot.current.removeChild(el.current);
      // Not sure about modal root, though. Why do we need it even? Shouldn't we clean it up?
    };
  }, []);

  return el.current;
}

function Modal(props) {
  const ariaId = React.useRef(uuid());
  const $el = useModalDiv();
  const portalRef = React.ureRef();

  const { onEscKey } = props;

  const handleKeyDown = React.useCallback(
    event => {
      if (event.key === "Escape") {
        onEscKey();
      }
    },
    [onEscKey]
  );

  useEscHandler(handleKeyDown);

  const setRef = React.useCallback(
    node => {
      if (node !== null) {
        portalRef.current = node.getBoundingClientRect;
        this.props.portalRef(node);
      }
    },
    [portalRef]
  );

  return ReactDOM.createPortal(
    () => (
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      <div
        className={classNames(className, { "modal--is-open": props.isOpen })}
        css={modalStyles}
        isOpen={props.isOpen}
        width={props.width}
        onKeyDown={handleKeyDown}
        ref={setRef}
        zIndex={props.zIndex}
      >
        {props.hasOverlay ? (
          <div
            css={props.overlayStyles}
            onClick={props.onOverlayClick}
            onKeyDown={props.onOverlayClick}
            tabIndex="-1"
          />
        ) : null}
        <div css={frameStyles} width={props.width}>
          <section
            css={contentStyles}
            isOpen={props.isOpen}
            isScrollable={props.isScrollable}
            role="dialog"
            aria-labelledby={this.ariaId}
          >
            <span css={stylers.visuallyHidden} id={this.ariaId}>
              {props.ariaLabel}
            </span>
            {props.children}
          </section>
        </div>
      </div>
    ),
    $el
  );
}

```

Reply:
@nahumzs @alexzherdev @rauldudu @mikrotron

Reply:
Just to give some context, the JS part is basically a copy/paste from acl-ui. Next step would be refactoring using hooks...

Reply:
Yep, save that snippet for round 2! Thanks, @AndreyChernykh!

Reply:
@mikrotron You're welcome!

Reply:
I just saw this, but is a great idea man!!